automates/model_analysis/graph_manipulation.py

```diff
@@ -0,0 +1,634 @@
+    :param g: graph (with named nodes)
+    :return: list of vertex names
+    """
+    name_list = g.vs["name"]
+    name_sorted = [name_list[i] for i in indices]
+    return name_sorted
+
+
+def wrap_d_sep(g, x, y, z):
+    """
+    Does some quick checks before testing d-separation
+    :param g: Graph
+    :param x: nodes
+    :param y: nodes
+    :param z: nodes
+    :return: T/F if x is separated from y given z in g
+    """
+    if x == y:
+        return False
+    if len(x) == 0 or len(y) == 0:
+        return True
+    return d_sep(g, x, y, z)
+
+
+def d_sep(g, x, y, z):
+    """
+    From R package causaleffect:
+
+    "Implements relevant path separation (rp-separation) for testing d-separation. For details, see:
+
+    Relevant Path Separation: A Faster Method for Testing Independencies in Bayesian Networks
+    Cory J. Butz, Andre E. dos Santos, Jhonatan S. Oliveira;
+    Proceedings of the Eighth International Conference on Probabilistic Graphical Models,
+    PMLR 52:74-85, 2016."
+
+    :param g: graph
+    :param x: nodes
+    :param y: nodes
+    :param z: nodes
+    :return: T/F if x is separated from y given z in g
+    """
+    an_z = ancestors_unsort(z, g)
+    an_xyz = ancestors_unsort(list(set(x) | set(y) | set(z)), g)
+    stack_top = len(x)
+    stack_size = max(stack_top, 64)
+    stack = [False] * stack_size
+    stack[0:stack_top] = [True] * len(range(0, stack_top))
+    stack_names = [None] * stack_size
+    stack_names[0:stack_top] = copy.deepcopy(x)
+    visited_top = 0
+    visited_size = 64
+    visited = [False] * visited_size
+    visited_names = [None] * visited_size
+    is_visited = False
+    while stack_top > 0:
+        is_visited = False
+        el = stack[stack_top - 1]
+        el_name = stack_names[stack_top - 1]
+        stack_top = stack_top - 1
+        if visited_top > 0:
+            for i in range(0, visited_top):
+                if el == visited[i] and el_name == visited_names[i]:
+                    is_visited = True
+                    break
+        if not is_visited:
+            if el_name in y:
+                return False
+            visited_top = visited_top + 1
+            if visited_top > visited_size:
+                visited_old = copy.deepcopy(visited)
+                visited_size_old = visited_size
+                visited_names_old = copy.deepcopy(visited_names)
+                visited_size = 2 * visited_size
+                visited = [False] * visited_size
+                visited[0:visited_size_old] = copy.deepcopy(visited_old)
+                visited_names = [None] * visited_size
+                visited_names[0:visited_size_old] = copy.deepcopy(visited_names_old)
+            visited[visited_top - 1] = el
+            visited_names[visited_top - 1] = el_name
+            el_name_in_z = el_name in z
+            if el and (not el_name_in_z):
+                visitable_parents = list((set(parents_unsort([el_name], g)) - set([el_name])) & set(an_xyz))
```

Comment:
@aswinchester the code from line 417 -- 435 seems very similar to the code from line 438 -- 452. In fact, it seems the single difference rests on whether "parents" are included in the various calculations or not. Why don't you turn this code into a Python function scoped under the current function with an arg that controls whether "parents" are used or not and then cleverly design the singular function to perform the roles of both of the code blocks I outlined above. Here is an example of how to scope functions under functions in Python:
```python
def foo(a, b, parent_flag=False):
    def foo_helper():
        if not parent_flag:
            ...
       ...
```
Notice how functions scoped under a parent scope still have access to all of the variables in the parent scope. This should be helpful to you in designing your function for these two code blocks. Not only is this a good programming exercise but it will also increase the maintainability of your code. If you ever need to change these code blocks you will only have to do it in one place instead of two 🙂.

BTW, you can treat function names as variables in Python so it is perfectly acceptable for you to send the function names `children_unsort` and `parent_unsort` as values into a function. That may also be helpful for you when designing this function for your two blocks.

Reply:
I wrote a helper function that is called three times in d-sep. This should address this and the following suggestion.